src/sleep.rs

```diff
@@ -0,0 +1,376 @@
+    pub fn new(pin: AnyInputPin, pullup: bool, pulldown: bool) -> Self {
+        Self {
+            pin,
+            pullup,
+            pulldown,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl fmt::Debug for RtcWakeupPin {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "RtcWakeupPin {{ pin: {:?} pullup {:?} pulldown {:?}}}",
+            self.pin.pin(),
+            self.pullup,
+            self.pulldown
+        )
+    }
+}
+
+/// Will wake up the CPU based on changes in GPIO pins. Is only available for light sleep.
+/// It takes PinDriver as input, which means that any required configurations on the pin
+/// can be done before adding it to the GpioWakeup struct.
+#[derive(Debug)]
+pub struct GpioWakeup<'a> {
+    pub pins: &'a [GpioWakeupPin<'a>],
+}
+
+impl<'a> GpioWakeup<'a> {
+    pub fn new(pins: &'a [GpioWakeupPin<'a>]) -> Self {
+        Self { pins }
+    }
+}
+
+impl<'a> GpioWakeup<'a> {
+    pub fn apply(&self) -> Result<(), EspError> {
+        for pin in self.pins.iter() {
+            let intr_level = match pin.wake_level {
+                Level::Low => gpio_int_type_t_GPIO_INTR_LOW_LEVEL,
+                Level::High => gpio_int_type_t_GPIO_INTR_HIGH_LEVEL,
+            };
+            esp!(unsafe { gpio_wakeup_enable(pin.pin.pin(), intr_level) })?;
+        }
+        esp!(unsafe { esp_sleep_enable_gpio_wakeup() })?;
+        Ok(())
+    }
+}
+
+pub struct GpioWakeupPin<'a> {
+    pub pin: PinDriver<'a, AnyInputPin, Input>,
+    pub wake_level: Level,
+}
+
+impl<'a> GpioWakeupPin<'a> {
+    pub fn new(
+        pin: PinDriver<'a, AnyInputPin, Input>,
+        wake_level: Level,
+    ) -> Result<Self, EspError> {
+        Ok(Self { pin, wake_level })
+    }
+}
+
+impl<'a> fmt::Debug for GpioWakeupPin<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "GpioWakeupPin {{ pin: {:?} wake_level: {:?} }}",
+            self.pin.pin(),
+            self.wake_level
+        )
+    }
+}
+
+/// Will wake up the CPU when the number of positive edges higher than a threshold
+/// is detected on the UART RX pin. Is only available for light sleep.
+pub struct UartWakeup<'a> {
+    pub uart: &'a UartDriver<'a>,
+    pub threshold: i32,
+}
+
+impl<'a> UartWakeup<'a> {
+    pub fn new(uart: &'a UartDriver<'a>, threshold: i32) -> Self {
+        Self { uart, threshold }
+    }
+
+    pub fn apply(&self) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_wakeup_threshold(self.uart.port(), self.threshold) })?;
+        esp!(unsafe { esp_sleep_enable_uart_wakeup(self.uart.port()) })?;
+        Ok(())
+    }
+}
+
+impl<'a> fmt::Debug for UartWakeup<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "UartWakeup")
+    }
+}
+
+/// Will wake up the CPU when a touchpad is touched.
+#[cfg(any(esp32, esp32s2, esp32s3))]
+#[derive(Debug)]
+pub struct TouchWakeup {}
```

Comment:
What is unclear to me here is if the touch pins are not configured to work in touch mode, whether wakeup would even work?

Reply:
`esp_sleep_enable_touchpad_wakeup()` will fail if you run it without setting up touch, just as `esp_sleep_enable_ulp_wakeup()` will fail if ULP is not running (probably needs to be set up to actually wake up as well, but not sure)

So, since neither of these takes any arguments, just assumes that things are set up, I think that at least as a starting point this lib can do the same, i.e. just "blindly" apply and check return value. 

Maybe this can be improved upon later, hopefully by somebody with abit better understanding and experience with touch/ulp. Maybe ill play around with that at some point, but not any time soon.